File: src/literature-web/src/pdfDownload.ts
import { api, Article, request, requestBlob, sessionGeneration, SourceOutcome } from "./api";
import { PdfRequest } from "./pdfRequest";
import { PlanPage, PlanSummary, validatePage } from "./plans/api";
import { exportPlan } from "./plans/exports";
import { structuredExport } from "./structuredExport";
import { TransferOptions } from "./transfer";
import { originalKind } from "./originalKind";

export type PdfResult = {
  id: string; article: Article; ready: boolean; pending: boolean; reason: string;
  hash?: string; bytes?: number; format?: string; mediaType?: string; outcome?: SourceOutcome;
};
export type PdfStatus = { items: PdfResult[]; ready: number; pending: number; unresolved: number; plan?: PlanSummary };
type PdfFile = { blob: Blob | null; filename: string; status: PdfStatus };
const invalid = () => new Error("The PDF response did not match this selection. No file was downloaded.");
export function pdfCurrent(intent: PdfRequest, signal: AbortSignal) {
  if (signal.aborted || intent.generation !== sessionGeneration()) throw new Error("PDF download context changed.");
}

// Reading progress never admits work or retries an acquisition. Every returned
// member must belong to the immutable selection captured by this user intent.
export async function readPdfStatus(intent: PdfRequest, signal: AbortSignal): Promise<PdfStatus> {
  pdfCurrent(intent, signal);
  const id = intent.confirmedID;
  if (!id) throw invalid();
  let items: PdfResult[], plan: PlanSummary | undefined;
  if (intent.kind === "batch") {
    const page = await api.batch(intent.library, id, intent.generation, signal);
    if (page.batch?.batch_id !== id || page.requestedFormat !== "pdf" || !Array.isArray(page.items) || page.total !== intent.body.searchIDs.length) throw invalid();
    items = page.items.map(i => ({ id: i.search_id, article: i.article, ready: i.downloadAvailable,
      pending: !["paused", "cancelled"].includes(page.batch.state) && ["queued", "running"].includes(i.state),
      reason: i.reason ?? "", hash: i.original_hash, bytes: i.bytes, format: i.format, mediaType: i.mediaType, outcome: i.sourceOutcome }));
  } else {
    const page = validatePage(await request<PlanPage>(`/api/libraries/${encodeURIComponent(intent.library)}/plans/${encodeURIComponent(id)}?offset=0&limit=100`, { signal }, intent.generation, 8 * 1024 * 1024));
    plan = page.plan;
    if (plan.planID !== id || plan.runID !== intent.runID || plan.scopeKind || plan.requestedFormat !== "pdf" || page.offset !== 0 || page.limit !== 100 || plan.selectedCount !== intent.body.searchIDs.length) throw invalid();
    items = page.items.map(i => ({ id: i.searchID, article: i.article, ready: i.downloadAvailable,
      pending: !["paused", "cancelled"].includes(plan!.state) && ["waiting", "queued", "running"].includes(i.phase),
      reason: i.reason, hash: i.original_hash, bytes: i.bytes, format: i.format, mediaType: i.mediaType, outcome: i.sourceOutcome }));
  }
  pdfCurrent(intent, signal);
  const selected = new Set(intent.body.searchIDs);
  if (items.length !== selected.size || new Set(items.map(i => i.id)).size !== selected.size ||
      items.some(i => !selected.has(i.id) || typeof i.ready !== "boolean" || !i.article || i.article.SearchId !== i.id ||
        (i.ready && (i.pending || originalKind({ downloadAvailable: i.ready, original_hash: i.hash, format: i.format, mediaType: i.mediaType }) !== "pdf" || !/^[a-f0-9]{64}$/.test(i.hash ?? "") || !Number.isSafeInteger(i.bytes) || i.bytes! < 1 || i.bytes! > 8 * 1024 * 1024)))) throw invalid();
  const ready = items.filter(i => i.ready).length, pending = items.filter(i => i.pending).length;
  return { items, ready, pending, unresolved: items.length - ready - pending, plan };
}

export async function pdfFile(intent: PdfRequest, status: PdfStatus, signal: AbortSignal, transfer: TransferOptions = {}, itemID?: string): Promise<PdfFile> {
  pdfCurrent(intent, signal);
  const item = itemID ? status.items.find(i => i.id === itemID) : status.items.length === 1 ? status.items[0] : undefined;
  let blob: Blob, filename: string;
  if (itemID && !item) throw invalid();
  if (item) {
    if (originalKind({ downloadAvailable: item.ready, original_hash: item.hash, format: item.format, mediaType: item.mediaType }) !== "pdf" || !item.hash) throw new Error("This original PDF is not currently available. Open its source links or check again.");
    blob = await requestBlob(`/api/libraries/${encodeURIComponent(intent.library)}/originals/${encodeURIComponent(item.id)}/${encodeURIComponent(item.hash)}`,
      { signal }, intent.generation, { ...transfer, maxBytes: 8 * 1024 * 1024 });
    pdfCurrent(intent, signal);
    const bytes = await blob.arrayBuffer(); pdfCurrent(intent, signal);
    const digest = await crypto.subtle.digest("SHA-256", bytes); pdfCurrent(intent, signal);
    const hash = Array.from(new Uint8Array(digest), b => b.toString(16).padStart(2, "0")).join("");
    if (blob.type.split(";")[0].trim().toLowerCase() !== "application/pdf" || blob.size !== item.bytes || hash !== item.hash ||
        new TextDecoder().decode(bytes.slice(0, 5)) !== "%PDF-") throw new Error("The original PDF failed its integrity check. No file was downloaded.");
    filename = `litradock-${item.id.replace(/[^a-zA-Z0-9_-]/g, "_")}.pdf`;
  } else {
    if (!status.ready) throw new Error("No original PDF is available for this selection.");
    return pdfPackageFile(intent, status, signal, transfer);
  }
  pdfCurrent(intent, signal);
  return { blob, filename, status };
}

type PackageItem = { searchId: string; available: boolean; file?: string; sha256?: string; bytes?: number; sourceOutcome: SourceOutcome };
// A versioned bounded response binds final scope and current outcomes to the
// exact archive hash. No earlier ready count is evidence of ZIP membership.
async function pdfPackageFile(intent: PdfRequest, previous: PdfStatus, signal: AbortSignal, transfer: TransferOptions): Promise<PdfFile> {
  const path = `/api/libraries/${encodeURIComponent(intent.library)}` + (intent.kind === "plan" ? `/plans/${encodeURIComponent(intent.confirmedID!)}/exports` : "/exports");
  const body = intent.kind === "plan" ? { format: "pdf-download" } : { batchID: intent.confirmedID, format: "pdf-download" };
  const response = await requestBlob(path, { method: "POST", body: JSON.stringify(body), signal }, intent.generation, { ...transfer, maxBytes: 38 * 1024 * 1024 + 4 });
  pdfCurrent(intent, signal);
  if (response.type.split(";")[0].trim().toLowerCase() !== "application/vnd.litradock.pdf-download" || response.size < 4) throw invalid();
  const prefix = await response.slice(0, 4).arrayBuffer(); pdfCurrent(intent, signal);
  const length = new DataView(prefix).getUint32(0);
  if (length < 2 || length > 1024 * 1024 || 4 + length > response.size) throw invalid();
  let report;
  try { report = JSON.parse(await response.slice(4, 4 + length).text()); } catch { throw invalid(); }
  pdfCurrent(intent, signal);
  if (!report || report.schema !== "litradock.pdf-download" || report.schemaVersion !== 1 || report.kind !== intent.kind || report.id !== intent.confirmedID ||
      report.requestedFormat !== "pdf" || (intent.kind === "plan" && report.runID !== intent.runID) || report.selectedCount !== intent.body.searchIDs.length ||
      !Array.isArray(report.items) || report.items.length !== report.selectedCount || !Number.isSafeInteger(report.includedRecords) ||
      report.includedRecords < 0 || report.includedRecords > report.selectedCount || report.unresolvedRecords !== report.selectedCount - report.includedRecords) throw invalid();
  const byID = new Map(previous.items.map(i => [i.id, i])), seen = new Set<string>();
  const items = report.items.map((value: PackageItem): PdfResult => {
    const old = value && byID.get(value.searchId), o = value?.sourceOutcome;
    if (!old || seen.has(value.searchId) || typeof value.available !== "boolean" || !o || o.requestedFormat !== "pdf" ||
        typeof o.status !== "string" || typeof o.label !== "string" || typeof o.detail !== "string" || typeof o.nextAction !== "string" ||
        (o.status === "ready") !== value.available) throw invalid();
    seen.add(value.searchId);
    if (value.available && (!/^[a-f0-9]{64}$/.test(value.sha256 ?? "") || !Number.isSafeInteger(value.bytes) || value.bytes! < 1 || value.bytes! > 8 * 1024 * 1024 ||
        value.file !== (intent.kind === "plan" ? `originals/${value.sha256}.pdf` : `originals/${value.searchId}-${value.sha256}.pdf`) ||
        (old.ready && (old.hash !== value.sha256 || old.bytes !== value.bytes)))) throw invalid();
    if (!value.available && (value.file || value.sha256 || value.bytes)) throw invalid();
    return { ...old, ready: value.available, pending: !value.available && ["queued", "running", "waiting"].includes(o.status),
      reason: o.detail, outcome: o, hash: value.sha256, bytes: value.bytes, format: value.available ? "PDF" : "", mediaType: value.available ? "application/pdf" : "" };
  });
  const ready = items.filter((i: PdfResult) => i.ready).length, pending = items.filter((i: PdfResult) => i.pending).length;
  if (ready !== report.includedRecords || seen.size !== intent.body.searchIDs.length || intent.body.searchIDs.some(id => !seen.has(id))) throw invalid();
  const status: PdfStatus = { items, ready, pending, unresolved: items.length - ready - pending, plan: previous.plan };
  const filename = "litradock-pdfs-and-outcomes.zip";
  const archive = response.slice(4 + length, response.size, "application/zip");
  if (!ready) {
    if (archive.size !== 0 || report.archiveBytes !== 0 || report.archiveSha256 !== "") throw invalid();
    return { blob: null, filename, status };
  }
  if (!/^[a-f0-9]{64}$/.test(report.archiveSha256) || archive.size !== report.archiveBytes || archive.size < 4 || archive.size > 37 * 1024 * 1024) throw invalid();
  const data = await archive.arrayBuffer(); pdfCurrent(intent, signal);
  const hash = await crypto.subtle.digest("SHA-256", data); pdfCurrent(intent, signal);
  if (Array.from(new Uint8Array(hash), b => b.toString(16).padStart(2, "0")).join("") !== report.archiveSha256 || new Uint8Array(data.slice(0, 4)).join(",") !== "80,75,3,4") throw invalid();
  return { blob: archive, filename, status };
}

export async function pdfOutcomes(intent: PdfRequest, status: PdfStatus, signal: AbortSignal) {
  pdfCurrent(intent, signal);
  return intent.kind === "plan"
    ? exportPlan(intent.library, status.plan!, "json", intent.generation, signal)
    : structuredExport(intent.library, { batchID: intent.confirmedID! }, "json", intent.generation, signal);
}

// A browser does not expose whether the user ultimately saved the file. Only
// claim a handoff, and retain an explicit Save again action for a blocked one.
export function handoffPdf(file: { blob: Blob; filename: string }, intent: PdfRequest, signal: AbortSignal) {
  pdfCurrent(intent, signal);
  const url = URL.createObjectURL(file.blob), anchor = document.createElement("a");
  try {
    anchor.href = url; anchor.download = file.filename; anchor.hidden = true;
    document.body.append(anchor); pdfCurrent(intent, signal); anchor.click();
  } finally { anchor.remove(); setTimeout(() => URL.revokeObjectURL(url), 60_000); }
}

export function waitForPdf(signal: AbortSignal) {
  return new Promise<void>((resolve, reject) => {
    const abort = () => { clearTimeout(timer); signal.removeEventListener("abort", abort); reject(new Error("Checking stopped.")); };
    const timer = setTimeout(() => { signal.removeEventListener("abort", abort); resolve(); }, 2000);
    signal.addEventListener("abort", abort, { once: true });
    if (signal.aborted) abort();
  });
}

File: src/literature-server/pdf_download.go
package main

import (
	"archive/zip"
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"io"
	"net/http"
)

const pdfDownloadMedia = "application/vnd.litradock.pdf-download"
const pdfReportBytes = 1024 * 1024

type pdfDownloadItem struct {
	SearchID  string         `json:"searchId"`
	Available bool           `json:"available"`
	File      string         `json:"file,omitempty"`
	SHA256    string         `json:"sha256,omitempty"`
	Bytes     int64          `json:"bytes,omitempty"`
	Outcome   *sourceOutcome `json:"sourceOutcome"`
}

type pdfDownloadReport struct {
	Schema       string            `json:"schema"`
	Version      int               `json:"schemaVersion"`
	Kind         string            `json:"kind"`
	ID           string            `json:"id"`
	RunID        string            `json:"runID,omitempty"`
	Format       string            `json:"requestedFormat"`
	Selected     int               `json:"selectedCount"`
	Included     int               `json:"includedRecords"`
	Unresolved   int               `json:"unresolvedRecords"`
	ArchiveBytes int               `json:"archiveBytes"`
	ArchiveHash  string            `json:"archiveSha256"`
	Items        []pdfDownloadItem `json:"items"`
}

// The primary PDF action consumes the final archive's manifest, never an earlier
// ready count. Ordinary ZIP exports remain useful even with metadata alone.
// Framing is a four-byte big-endian JSON length, bounded JSON, then exact ZIP
// bytes; zero included PDFs returns only the outcome report and no archive.
func pdfDownloadPackage(archive []byte, kind, id string) ([]byte, []byte, error) {
	bad := func() ([]byte, []byte, error) {
		return nil, nil, exportInvalid("PDF package scope or integrity mismatch")
	}
	if len(archive) > planZIPBytes {
		return bad()
	}
	z, err := zip.NewReader(bytes.NewReader(archive), int64(len(archive)))
	if err != nil || len(z.File) > 102 {
		return bad()
	}
	files := map[string]*zip.File{}
	for _, f := range z.File {
		if files[f.Name] != nil {
			return bad()
		}
		files[f.Name] = f
	}
	f := files["manifest.json"]
	if f == nil || f.UncompressedSize64 > 8*1024*1024 {
		return bad()
	}
	r, err := f.Open()
	if err != nil {
		return bad()
	}
	manifest, err := io.ReadAll(io.LimitReader(r, 8*1024*1024+1))
	r.Close()
	if err != nil || len(manifest) > 8*1024*1024 {
		return bad()
	}
	d := pdfDownloadReport{Schema: "litradock.pdf-download", Version: 1, Kind: kind, ID: id, Format: "pdf", Items: []pdfDownloadItem{}}
	metadata := "records.csv"
	if kind == "batch" {
		var m struct {
			ID     string `json:"batchId"`
			Format string `json:"requestedFormat"`
			Items  []struct {
				SearchID string         `json:"searchId"`
				File     string         `json:"file"`
				SHA256   string         `json:"sha256"`
				Bytes    int64          `json:"bytes"`
				Format   string         `json:"format"`
				Media    string         `json:"mediaType"`
				Outcome  *sourceOutcome `json:"sourceOutcome"`
			} `json:"items"`
		}
		if json.Unmarshal(manifest, &m) != nil || m.ID != id || m.Format != "pdf" {
			return bad()
		}
		for _, i := range m.Items {
			if i.File != "" && (i.Format != "PDF" || i.Media != "application/pdf" || i.File != "originals/"+i.SearchID+"-"+i.SHA256+".pdf") {
				return bad()
			}
			d.Items = append(d.Items, pdfDownloadItem{i.SearchID, i.File != "", i.File, i.SHA256, i.Bytes, i.Outcome})
		}
	} else if kind == "plan" {
		var m planExportDocument
		if json.Unmarshal(manifest, &m) != nil || m.Plan.PlanID != id || m.Plan.RequestedFormat != "pdf" || m.Plan.ScopeKind != "" || !m.Revalidated || m.Plan.SelectedCount != len(m.Items) {
			return bad()
		}
		d.RunID = m.Plan.RunID
		metadata = "records.json"
		for _, i := range m.Items {
			entry := pdfDownloadItem{SearchID: i.SearchID, Available: i.Availability == "included", Outcome: i.SourceOutcome}
			if entry.Available {
				if i.Original == nil || i.File == nil || i.Original.Format != "PDF" || i.Original.MediaType != "application/pdf" || *i.File != "originals/"+i.Original.SHA256+".pdf" {
					return bad()
				}
				entry.File, entry.SHA256, entry.Bytes = *i.File, i.Original.SHA256, i.Original.Bytes
			} else if i.File != nil || i.Original != nil {
				return bad()
			}
			d.Items = append(d.Items, entry)
		}
	} else {
		return bad()
	}
	if len(d.Items) < 1 || len(d.Items) > 100 || files[metadata] == nil || files[metadata].UncompressedSize64 > 8*1024*1024 {
		return bad()
	}
	used := map[string]bool{"manifest.json": true, metadata: true}
	selected := map[string]bool{}
	var total int64
	for _, i := range d.Items {
		if !savedIDPattern.MatchString(i.SearchID) || selected[i.SearchID] || i.Outcome == nil || i.Outcome.RequestedFormat != "pdf" || (i.Outcome.Status == "ready") != i.Available {
			return bad()
		}
		selected[i.SearchID] = true
		if !i.Available {
			continue
		}
		d.Included++
		f := files[i.File]
		expected, e := hex.DecodeString(i.SHA256)
		if e != nil || len(expected) != 32 || i.Bytes < 1 || i.Bytes > 8*1024*1024 || f == nil || f.UncompressedSize64 != uint64(i.Bytes) || f.Method != zip.Store {
			return bad()
		}
		if used[i.File] {
			continue
		}
		used[i.File] = true
		total += i.Bytes
		if total > planOriginalBytes {
			return bad()
		}
		r, e := f.Open()
		if e != nil {
			return bad()
		}
		h := sha256.New()
		prefix := make([]byte, 5)
		_, e = io.ReadFull(r, prefix)
		if e != nil || string(prefix) != "%PDF-" {
			r.Close()
			return bad()
		}
		h.Write(prefix)
		n, e := io.Copy(h, io.LimitReader(r, 8*1024*1024))
		r.Close()
		if e != nil || n+5 != i.Bytes || !bytes.Equal(h.Sum(nil), expected) {
			return bad()
		}
	}
	if len(files) != len(used) {
		return bad()
	}
	d.Selected = len(d.Items)
	d.Unresolved = d.Selected - d.Included
	if d.Included == 0 {
		archive = nil
	} else {
		d.ArchiveBytes = len(archive)
		hash := sha256.Sum256(archive)
		d.ArchiveHash = hex.EncodeToString(hash[:])
	}
	header, err := json.Marshal(d)
	if err != nil || len(header) > pdfReportBytes {
		return bad()
	}
	return header, archive, nil
}

func writePDFDownload(w http.ResponseWriter, report, archive []byte) {
	w.Header().Set("Content-Type", pdfDownloadMedia)
	var length [4]byte
	binary.BigEndian.PutUint32(length[:], uint32(len(report)))
	if _, err := w.Write(length[:]); err != nil {
		return
	}
	if _, err := w.Write(report); err != nil {
		return
	}
	_, _ = w.Write(archive)
}
